test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+        const (m,n) = rhs.shape;
+        if m != 1 then
+            err("Trying to apply a vector of shape ",lhs.shape, " to a matrix of shape ", rhs.shape, ". m needs to be 1");
+        
+        var b = new Tensor(rank=2,eltType=eltType);
+        b.reshapeDomain({0..#p, 0..#n});
+        foreach (i,j) in {0..#p, 0..#n} {
+            b[i,j] = lhs[i] * rhs[0,j];
+        }
+        return b; 
+    }
+
+    // Matrix-matrix multiplication
+    operator *(lhs: Tensor(2,?eltType), rhs: Tensor(2,eltType)): Tensor(2,eltType) {
+        const (m,n) = lhs.shape;
+        const (p,q) = rhs.shape;
+        if n != p then
+            err("Trying to apply a matrix of shape ",lhs.shape, " to a matrix of shape ", rhs.shape);
+
+        const a = lhs.data;
+        const b = rhs.data;
+        var c = new Tensor(rank=2,eltType=eltType);
+        c.reshapeDomain({0..#m, 0..#q});
+        forall (i,j) in c.domain with (ref c) {
+            const row = a[i,..];
+            const col = b[..,j];
+            c[i,j] = + reduce (row * col);
+        }
+        return c;
+    }
+
+    operator /(lhs: Tensor(?d,?eltType), c: eltType) {
+        const data = lhs.data / c;
+        return new Tensor(data);
+    }
+    operator -(lhs: Tensor(?d,?eltType), c: eltType) {
+        const data = lhs.data - c;
+        return new Tensor(data);
+    }
+
+    // Sigmoid function
+    proc _sigmoid(x: real): real {
+        return 1.0 / (1.0 + Math.exp(-x));
+    }
+    // Derivative of sigmoid function
+    proc _sigmoidPrime(x: real): real {
+        const s = _sigmoid(x);
+        return s * (1.0 - s);
+    }
+
+    // Apply sigmoid function to each element of tensor
+    proc sigmoid(t: Tensor(?d)): Tensor(d) {
+        return t.fmap(_sigmoid);
+    }
+
+    // Apply derivative of sigmoid function to each element of tensor
+    proc sigmoidPrime(t: Tensor(?d)): Tensor(d) {
+        return t.fmap(_sigmoidPrime);
+    }
+
+    // Get forbenius norm before sqrt
+    proc frobeniusNormPowTwo(t: Tensor(?d)): real {
+        const AA = t.data ** 2.0;
+        return + reduce AA;
+    }
+
+    // Apply exponential function to each element of tensor
+    proc exp(t: Tensor(?d)): Tensor(d) {
+        var y = new Tensor(t.domain);
+        foreach i in t.domain do
+            y.data[i] = Math.exp(t.data[i]);
+        return y;
+    }
+
+    // Wikipedia implementation (helper for randn)
+    // mu : mean
+    // sigma : standard deviation
+    proc boxMuller(mu: real, sigma: real) {
+        var u1 = rng.getNext();
+        var u2 = rng.getNext();
+        var z0 = sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.pi * u2);
+        return mu + (sigma * z0);
+    }
+    proc normal() {
+        return boxMuller(0.0,1.0);
+    }
+
+    // Initialize a tensor with random values from a normal distribution
+    proc randn(shape: int ...?d): Tensor(d,real) {
+        var t = new Tensor((...shape));
+        for i in t.domain {
+            t.data[i] = normal();
```

Comment:
Instead of computing one bux-muller value at a time, I'd probably do the whole tensor in bulk with the `fillRandom` method on the rng. Something like:
```chapel
var u1, u2: [domainFromShape((...shape))] real;
fillRandom(u1);
fillRandom(u2);
const u1p = -2.0 * Math.log(u1),
      u2p = 2.0 * Math.pi * u2;
var z0 = sqrt(u1p);
z0 *= Math.cos(u2p);
return new Tensor(z0);
```

Reply:
Does fillRandom sample from a normal distribution?

Reply:
I don't think so, I'm pretty sure it uses a uniform distribution.

Reply:
The network initialization requires a normal distribution, so that is why this is there. I really don't understand how the variance of the elements in a matrix effects training, but this is what I found in "Deep Dive into Machine Learning"

Reply:
Right, the code I pasted above starts with `u1` and `u2` which are uniformly distributed values, and then applies the Box-muller algorithm to them to produce a normally distributed array.

Reply:
Oh I see. I made the change.